horizon/enforcer/utils/mapping_rules_utils.py

```diff
@@ -102,25 +104,84 @@ def extract_attributes_from_query_params(rule_url: str, request_url: str) -> dic
                 attributes[rule_query_params[key][1:-1]] = request_query_params[key]
         return attributes
 
+    @classmethod
+    def _compare_urls(
+        cls, 
+        mapping_rule_url: str, 
+        request_url: str, 
+        is_regex: bool = False
+    ) -> bool:
+        """
+        Compare a mapping rule URL against a request URL.
+        """
+        # If the mapping rule is a regex pattern
+        if is_regex:
+            try:
+                pattern = re.compile(mapping_rule_url)
+                match_result = bool(pattern.match(request_url))
+                logger.debug(
+                    "regex url comparison",
+                    pattern=mapping_rule_url,
+                    url=request_url,
+                    matched=match_result
+                )
+                return match_result
+            except re.error as e:
+                logger.warning(
+                    "regex pattern compilation failed",
+                    pattern=mapping_rule_url,
+                    error=str(e)
+                )
+                return False
+        
+        # For traditional URL matching
+        try:
+            # Compare paths
+            if not cls._compare_url_path(mapping_rule_url, request_url):
+                return False
+            
+            return True
+        except Exception as e:
+            logger.warning(
+                "URL comparison failed - verify URL format and structure",
+                mapping_url=mapping_rule_url,
+                request_url=request_url,
+                error_message=str(e),
+                error_type=type(e).__name__
+            )
+            return False
+
     @classmethod
     def extract_mapping_rule_by_request(
-        cls,
-        mapping_rules: list[MappingRuleData],
-        http_method: str,
-        url: AnyHttpUrl,
-    ) -> MappingRuleData | None:
-        matched_mapping_rules = []
+        cls, 
+        mapping_rules: List[MappingRuleData], 
+        http_method: str, 
+        url: str
+    ) -> Optional[MappingRuleData]:
+        """Extract matching mapping rule for the given request"""
+        http_method = http_method.lower()  # Convert once instead of in each iteration
+        
         for mapping_rule in mapping_rules:
-            if not mapping_rule.http_method == http_method.lower():
-                # if the method is not the same, we don't need to check the url
+            is_regex = getattr(mapping_rule, 'url_type', None) == "regex"
```

Comment:
According to the type annotation, this should be a pydantic instance, meaning it must have `url_type`. Why did you use `getattr()` instead?
 ```suggestion
            is_regex = mapping_tule.url_type == "regex"
```

Reply:
Good catch. Fixed.